src/passes/OptimizeInstructions.cpp

```diff
@@ -841,6 +757,66 @@ struct OptimizeInstructions
           }
         }
       }
+      if (auto* condition = select->condition->dynCast<Unary>()) {
+        if (condition->op == EqZInt32) {
+          // flip select to remove eqz, if we can reorder
+          EffectAnalyzer ifTrue(getPassOptions(), features, select->ifTrue);
+          EffectAnalyzer ifFalse(getPassOptions(), features, select->ifFalse);
+          if (!ifTrue.invalidates(ifFalse)) {
+            select->condition = condition->value;
+            std::swap(select->ifTrue, select->ifFalse);
+          }
+        }
+      }
+      if (auto* constTrue = select->ifTrue->dynCast<Const>()) {
+        if (auto* constFalse = select->ifFalse->dynCast<Const>()) {
+          Builder builder(*getModule());
+          bool is64bits = select->type == Type::i64;
+          auto extendIfNeeded = [&](Expression* expression) {
+            return is64bits ? builder.makeUnary(ExtendUInt32, expression)
+                            : expression;
+          };
+          if (select->type == Type::i32 || is64bits) {
+            if (constTrue->type == constFalse->type &&
+                constTrue->value.getInteger() == 1LL &&
+                constFalse->value.getInteger() == 0LL) {
+              auto* condition = select->condition;
+              if (Properties::emitsBoolean(condition)) {
+                // !x ? 1 : 0   ==>   !x
+                // x <=> y ? 1 : 0   ==>   x <=> y
+                return extendIfNeeded(condition);
+              } else {
+                // expr ? 1 : 0   ==>   expr != 0
+                return extendIfNeeded(builder.makeBinary(
+                  NeInt32, condition, builder.makeConst(Literal(0))));
```

Comment:
this can be just an EqZ EqZ, which is shorter I think? 2 bytes, while the current code is 1 byte for the NeInt32 and two for the const 0.

Reply:
Yes, good catch!

Reply:
I decide use `EqZ(EqZ(x))` only for `shrinkLevel > 0` due to `Ne(x, 0)` usually produce faster code (for interpreted runtimes). Wdyt?